### PR TITLE
feat (tcomp): refactor tcomp archetype folder and add poc archetype for bd

### PR DESCRIPTION
**Pull request by StefanLMNG**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows Talend standard
- [ ] Tests for the changes have been added (for bug fixes / features)

**What is the current behavior?** (You can also link to an open issue here)
Maven archetype doesn't respect the new TCOMP infra

**What is the new behavior?**
New maven archetype which respect the new TCOMP infra

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: components-archetype/src/main/resources/archetype-resources/pom.xml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  <artifactId>${artifactId}-parent</artifactId>
+  <packaging>pom</packaging>
+
+  <name>${componentClass} component family</name>
+
+  <modules>
+    <module>${artifactId}-runtime_XXX</module>
+    <module>${artifactId}-definition</module>
+  </modules>
+<!--
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/pom.xml]: Dead code.

File: components-archetype/src/main/resources/archetype-resources/__artifactId__-runtime_XXX/src/main/java/runtime_XXX/__componentClass__SourceOrSink.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    @Override
+    public Schema getEndpointSchema(RuntimeContainer container, String schemaName) throws IOException {
+        return null;
+    }
+
+    // TODO what kind of check we need to do
+    boolean doValidate (String name, ComponentProperties properties){
+        try {
+            //${componentClass}SourceOrSink mdbsos = new ${componentClass}SourceOrSink();
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/__artifactId__-runtime_XXX/src/main/java/runtime_XXX/__componentClass__SourceOrSink.java]: Add some comments

File: components-archetype/src/main/resources/archetype-resources/__artifactId__-runtime_XXX/src/main/java/runtime_XXX/__componentClass__SourceOrSink.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    public ValidationResult validate(RuntimeContainer container) {
+        try {
+            connect(container);
+        } catch (IOException ex) {
+            return new ValidationResult().setStatus(ValidationResult.Result.ERROR).setMessage(ex.getMessage());
+        }
+        return ValidationResult.OK;
+    }
+
+    protected MongoClient connect(RuntimeContainer container) throws IOException {
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/__artifactId__-runtime_XXX/src/main/java/runtime_XXX/__componentClass__SourceOrSink.java]: Typo.

File: components-archetype/src/main/resources/archetype-resources/__artifactId__-runtime_XXX/src/main/java/runtime_XXX/__componentClass__SourceOrSink.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        try {
+            return new MongoClient(new MongoClientURI(getUri()));
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/__artifactId__-runtime_XXX/src/main/java/runtime_XXX/__componentClass__SourceOrSink.java]: Typo.

File: components-archetype/src/main/resources/archetype-resources/__artifactId__-runtime_XXX/src/main/java/runtime_XXX/__componentClass__SourceOrSink.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    boolean doValidate (String name, ComponentProperties properties){
+        try {
+            //${componentClass}SourceOrSink mdbsos = new ${componentClass}SourceOrSink();
+        } catch (Exception e) {
+            return false;
+        }
+        return true;
+    }
+
+    // TODO check if
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/__artifactId__-runtime_XXX/src/main/java/runtime_XXX/__componentClass__SourceOrSink.java]: Remove the TODO here

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/100/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/100/)

File: components-archetype/src/main/resources/archetype-resources/__artifactId__-runtime_XXX/src/main/java/runtime_XXX/__componentClass__SourceOrSink.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        }
+    }
+
+    public ${componentClass}DatastoreProperties getProperties() {
+        return properties;
+    }
+
+    @Override
+    public List<NamedThing> getSchemaNames(RuntimeContainer container) throws IOException {
+        return null;
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/__artifactId__-runtime_XXX/src/main/java/runtime_XXX/__componentClass__SourceOrSink.java]: Need a comments.

File: components-archetype/src/main/resources/archetype-resources/__artifactId__-runtime_XXX/src/main/java/runtime_XXX/__componentClass__SourceOrSink.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    @Override
+    public Schema getEndpointSchema(RuntimeContainer container, String schemaName) throws IOException {
+        return null;
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/__artifactId__-runtime_XXX/src/main/java/runtime_XXX/__componentClass__SourceOrSink.java]: Need a comments.

File: components-archetype/src/main/resources/archetype-resources/__artifactId__-runtime_XXX/src/main/java/runtime_XXX/__componentClass__SourceOrSink.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    // TODO what kind of check we need to do
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/__artifactId__-runtime_XXX/src/main/java/runtime_XXX/__componentClass__SourceOrSink.java]: Remove TODO.

File: components-archetype/src/main/resources/archetype-resources/__artifactId__-runtime_XXX/src/main/java/runtime_XXX/__componentClass__SourceOrSink.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+import org.talend.daikon.NamedThing;
+import org.talend.daikon.properties.ValidationResult;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class ${componentClass}SourceOrSink implements SourceOrSink{
+
+    protected static final String DATABASE_GLOBALMAP_KEY = "database"; //FIXME only for Studio?
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/__artifactId__-runtime_XXX/src/main/java/runtime_XXX/__componentClass__SourceOrSink.java]: Remove FixMe

File: components-archetype/src/main/resources/archetype-resources/__artifactId__-runtime_XXX/pom.xml [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    <relativePath>../pom.xml</relativePath>
+   </parent>
+
+  <artifactId>${artifactId}-runtime_XXX</artifactId>
+  <packaging>pom</packaging>
+
+  <name>Runtime module</name>
+
+  <dependencies>
+        <dependency>
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/__artifactId__-runtime_XXX/pom.xml]: Remove unecessary dependencies.

File: components-archetype/src/main/resources/META-INF/maven/archetype-metadata.xml [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+        <fileSet filtered="true">
+            <directory>__artifactId__-definition</directory>
+            <includes>
+                <include>pom.xml</include>
+            </includes>
+        </fileSet>
+
+       
+
+        <!--<fileSet filtered="true" encoding="UTF-8">
```
Comment (pbailly) [inline components-archetype/src/main/resources/META-INF/maven/archetype-metadata.xml]: Dead code.

File: components-archetype/src/main/resources/META-INF/maven/archetype-metadata.xml [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+        <requiredProperty key="componentClass">
+            <defaultValue>${component-name.substring(0,1).toUpperCase()}${component-name.substring(1)}</defaultValue>
+        </requiredProperty>
+        <requiredProperty key="package">
+            <defaultValue>${groupId}</defaultValue>
+        </requiredProperty>
+        <requiredProperty key="packageTalend">
+            <defaultValue>org.talend.components</defaultValue>
+        </requiredProperty>
+        <!--<requiredProperty key="resource-name">
```
Comment (pbailly) [inline components-archetype/src/main/resources/META-INF/maven/archetype-metadata.xml]: Dead code.

File: components-archetype/pom.xml [hunk 13]
```diff
@@ -0,0 +1,39 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+<!-->
```
Comment (pbailly) [inline components-archetype/pom.xml]: Dead code.

File: components-archetype/src/main/resources/archetype-resources/pom.xml [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+	</parent>
+  
+  <modelVersion>4.0.0</modelVersion>
+  <artifactId>${artifactId}-parent</artifactId>
+  <packaging>pom</packaging>
+
+  <name>${componentClass} component family</name>
+
+  <modules>
+    <module>${artifactId}-runtime_XXX</module>
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/pom.xml]: runtime_XXX should be replace by a variable

File: components-archetype/src/main/resources/archetype-resources/__artifactId__-definition/src/main/java/__componentClass__DatasetDefinition.java [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+
+    public ${componentClass}DatasetProperties createProperties() {
+        return new ${componentClass}DatasetProperties(${componentClass}ComponentFamilyDefinition.NAME);
+    }
+
+    public RuntimeInfo getRuntimeInfo(${componentClass}DatasetProperties properties, Object ctx) {
+        return new RuntimeInfo() {
+
+            public List<URL> getMavenUrlDependencies() {
+                return null;
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/__artifactId__-definition/src/main/java/__componentClass__DatasetDefinition.java]: Add a TODO.
Add a description of the method to ease the comprehension.

File: components-archetype/src/main/resources/archetype-resources/__artifactId__-definition/src/main/java/__componentClass__DatasetDefinition.java [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+            }
+
+            public String getRuntimeClassName() {
+                return null;
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/__artifactId__-definition/src/main/java/__componentClass__DatasetDefinition.java]: Add a TODO.
Add a description of the method to ease the comprehension.

File: components-archetype/src/main/resources/archetype-resources/__artifactId__-definition/src/main/java/__componentClass__ComponentFamilyDefinition.java [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+ */
+@Component(name = Constants.COMPONENT_INSTALLER_PREFIX
+        + ${componentClass}ComponentFamilyDefinition.NAME, provide = ComponentInstaller.class)
+public class ${componentClass}ComponentFamilyDefinition extends AbstractComponentFamilyDefinition implements ComponentInstaller {
+
+    public static final String NAME = "${component-name}";
+
+    public ${componentClass}ComponentFamilyDefinition() {
+        super(NAME, new ${componentClass}DatastoreDefinition(), new ${componentClass}DatastoreDefinition(), new ${componentClass}InputDefinition(),
+                new ${componentClass}OutputDefinition());
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/__artifactId__-definition/src/main/java/__componentClass__ComponentFamilyDefinition.java]: The user can add mutliple definition here, we have to warn him.
Add a TODO.

File: components-archetype/src/main/resources/archetype-resources/__artifactId__-definition/src/main/java/__componentClass__DatasetProperties.java [hunk 18]
```diff
@@ -0,0 +1,10 @@
+package ${package};
+
+import ${packageTalend}.common.dataset.DatasetProperties;
+import org.talend.daikon.properties.PropertiesImpl;
+
+public class ${componentClass}DatasetProperties extends PropertiesImpl implements DatasetProperties {
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/__artifactId__-definition/src/main/java/__componentClass__DatasetProperties.java]: This is where the user can add properties, we need to provide a very extensive comments about what he can do and how.
Also add a TODO

File: components-archetype/src/main/resources/archetype-resources/__artifactId__-definition/src/main/java/__componentClass__DatasetProperties.java [hunk 19]
```diff
@@ -0,0 +1,10 @@
...
+    public ${componentClass}DatasetProperties(String name) {
+        super(name);
+    }
+}
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/__artifactId__-definition/src/main/java/__componentClass__DatasetProperties.java]: It maybe a good thing to add a note about the "refreshLayout" method.

**Comment by pteyssier** [comment]
Commit message not Talend standard (I didn't approve dataflow team norms on components repo): no JIRA id to track the modification + notion "feature or bug" is not discriminant enough to be first part and it's tmi, unnecessary at commit level: to be discussed later, no need to update your commit message now.

File: components-archetype/src/main/resources/archetype-resources/__artifactId__-definition/src/main/java/__componentClass__DatastoreDefinition.java [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+
+public class ${componentClass}DatastoreDefinition extends SimpleNamedThing implements DatastoreDefinition<${componentClass}DatastoreProperties> {
+
+    @Override
+    public ${componentClass}DatastoreProperties createProperties() {
+        return new ${componentClass}DatastoreProperties(getName());
+    }
+
+    @Override
+    public RuntimeInfo getRuntimeInfo(${componentClass}DatastoreProperties properties, Object ctx) {
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/__artifactId__-definition/src/main/java/__componentClass__DatastoreDefinition.java]: Add a TODO.
Add a description of the method to ease the comprehension.

File: components-archetype/src/main/resources/archetype-resources/__artifactId__-definition/src/main/java/__componentClass__DatastoreProperties.java [hunk 21]
```diff
@@ -0,0 +1,20 @@
+package ${package};
+
+import ${packageTalend}.api.component.runtime.RuntimeInfo;
+import ${packageTalend}.common.datastore.DatastoreDefinition;
+import org.talend.daikon.SimpleNamedThing;
+import org.talend.daikon.i18n.I18nMessages;
+import org.talend.daikon.properties.Properties;
+
+public class ${componentClass}DatastoreDefinition extends SimpleNamedThing implements DatastoreDefinition<${componentClass}DatastoreProperties> {
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/__artifactId__-definition/src/main/java/__componentClass__DatastoreProperties.java]: Wrong class, this is DatastoreProperties, not DatastoreDefinition

File: components-archetype/src/main/resources/archetype-resources/__artifactId__-definition/src/main/java/input/__componentClass__InputDefinition.java [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    @Override
+    public Class<? extends ComponentProperties> getPropertyClass() {
+        return ${componentClass}InputProperties.class;
+    }
+
+    public RuntimeInfo getRuntimeInfo(Properties properties, ConnectorTopology connectorTopology) {
+        ${componentClass}InputProperties prop = (${componentClass}InputProperties) properties;
+        // Depending on the version, return a RuntimeInfo corresponding to ${componentClass}Source.
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/__artifactId__-definition/src/main/java/input/__componentClass__InputDefinition.java]: Add TODO before the comments in order to be more visual by the user.

File: components-archetype/src/main/resources/archetype-resources/__artifactId__-definition/src/main/java/input/__componentClass__InputDefinition.java [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+        return null;
+    }
+
+    public Property[] getReturnProperties() {
+        return new Property[0];
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/__artifactId__-definition/src/main/java/input/__componentClass__InputDefinition.java]: Add comments to explain this method
Add a TODO

File: components-archetype/src/main/resources/archetype-resources/__artifactId__-definition/src/main/java/input/__componentClass__InputDefinition.java [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    public Set<ConnectorTopology> getSupportedConnectorTopologies() {
+        return null;
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/__artifactId__-definition/src/main/java/input/__componentClass__InputDefinition.java]: Add comments to explain this method
Add a TODO

File: components-archetype/src/main/resources/archetype-resources/__artifactId__-definition/src/main/java/input/__componentClass__InputProperties.java [hunk 25]
```diff
@@ -0,0 +1,10 @@
+package ${package}.input;
+
+import ${packageTalend}.api.properties.ComponentPropertiesImpl;
+
+public class ${componentClass}InputProperties extends ComponentPropertiesImpl {
+
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/__artifactId__-definition/src/main/java/input/__componentClass__InputProperties.java]: This is where the user can add properties, we need to provide a very extensive comments about what he can do and how.
Also add a TODO

File: components-archetype/src/main/resources/archetype-resources/__artifactId__-definition/src/main/java/input/__componentClass__InputProperties.java [hunk 26]
```diff
@@ -0,0 +1,10 @@
...
+    public ${componentClass}InputProperties(String name) {
+        super(name);
+    }
+}
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/__artifactId__-definition/src/main/java/input/__componentClass__InputProperties.java]: Add a note about the "refreshLayout" method.

File: components-archetype/src/main/resources/archetype-resources/__artifactId__-runtime_XXX/src/main/java/runtime_XXX/__componentClass__Sink.java [hunk 27]
```diff
@@ -0,0 +1,4 @@
+package ${package}.runtime_XXX;
+
+public class ${componentClass}Sink {
+}
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/__artifactId__-runtime_XXX/src/main/java/runtime_XXX/__componentClass__Sink.java]: Add comments explaining what is this method and how to code inside it.

File: components-archetype/src/main/resources/archetype-resources/__artifactId__-runtime_XXX/src/main/java/runtime_XXX/__componentClass__Source.java [hunk 28]
```diff
@@ -0,0 +1,4 @@
+package ${package}.runtime_XXX;
+
+public class ${componentClass}Source {
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/__artifactId__-runtime_XXX/src/main/java/runtime_XXX/__componentClass__Source.java]: Add comments explaining what is this method and how to code inside it.

File: components-archetype/src/main/resources/archetype-resources/__artifactId__-runtime_XXX/src/main/java/runtime_XXX/__componentClass__SourceOrSink.java [hunk 29]
```diff
@@ -0,0 +1,74 @@
...
+
+    protected transient ${componentClass}DatastoreProperties properties;
+
+    @Override
+    public ValidationResult initialize(RuntimeContainer container, ComponentProperties properties) {
+        this.properties = (${componentClass}DatastoreProperties) properties;
```
Comment (pbailly) [inline components-archetype/src/main/resources/archetype-resources/__artifactId__-runtime_XXX/src/main/java/runtime_XXX/__componentClass__SourceOrSink.java]: Add comments

**Comment by talend-glorieux** [comment]
@pteyssier Well there are not mention of Jira on the [README](https://github.com/Talend/components/blob/master/README.md) and [CONTRIBUTING](https://github.com/Talend/components/blob/master/CONTRIBUTING.md) files and I can't see any reference to a commit message format guideline. 
So I would say that the pull request fully complies with the contribution workflow proposed on the [CONTRIBUTING](https://github.com/Talend/components/blob/master/CONTRIBUTING.md) file.

**Comment by pteyssier** [comment]
@talend-glorieux "So I would say that the pull request fully complies with the contribution workflow proposed on the CONTRIBUTING file." I was not referring to this file but since you do: Please check the screenshot there. Point 1 of the pull request chapter. Still not compliant.
There is an on-boarding wiki pointing to the wiki specifying the standard at Talend. You surely missed it. However, I would agree to discuss about changing it.

**Comment by talend-glorieux** [comment]
@pteyssier Good point. Even if to me, a screenshot doesn't make a rule.
It would be a good thing to improve the on boarding experience on that part to make sure that outside and inside contributors have the same level of knowledge.
There is actually a set of [predefined templates for the CONTRIBUTING.md file and pull requests body](https://github.com/Talend/tools/tree/master/tools-root-github) on the tools repository. Feel free to borrow from those and/or submit a pull request with enhancements.

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/233/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/233/)

**Comment by sgandon** [comment]
@StefanLMNG this is not clear what this PR is about, is it about a new JMS component or the BD archetype definition ?
Please make this more clear and do not submit both features in the same PR.

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/237/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/237/)

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/246/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/246/)

**Comment by StefanLMNG** [comment]
This pull request will be canceled, please follow this new one: #268

**Pull request closed without merging**